specs/BrowserProcessExited.md

```diff
@@ -0,0 +1,322 @@
+version after a `NewBrowserVersionAvailable` event.
+
+This event is raised for both expected and unexpected browser process
+termination, after all resources, including the user data folder, used by the
+browser process (and related processes) have been released. The
+`ICoreWebView2BrowserProcessExitedEventArgs` interface lets app developers get
+the `BrowserProcessExitKind` so they can decide how to handle different exit
+kinds or bypass handling if an event handler for the `CoreWebView2`s
+`ProcessFailed` event (for `CoreWebView2ProcessFailedKind.BrowserProcessFailed`)
+is already registered. In case of a browser process crash, both
+`BrowserProcessExited` and `ProcessFailed` events are raised, but the order is
+not guaranteed.
+
+All `CoreWebView2Environment` objects across different app processes that use
+the same browser process receive this event when the browser process (and
+associated processes) exits. If the browser process (and therefore the user data
+folder) in use by the app process (through the `CoreWebView2Environment` options
+used) is shared with other processes, these processes need to coordinate to
+handle the potential race condition on the use of the resources. E.g., if one
+app process tries to clear the user data folder, while other tries to recreate
+its WebViews on crash.
+
+
+# Examples
+The following code snippets demonstrate how the `BrowserProcessExited` event can
+be used:
+
+## Win32 C++
+```cpp
+// Before closing the WebView, register a handler with code to run once the
+// browser process is terminated.
+EventRegistrationToken browserExitedEventToken;
+
+CHECK_FAILURE(m_webViewEnvironment->add_BrowserProcessExited(
+    Callback<ICoreWebView2BrowserProcessExitedEventHandler>(
+        [browserExitedEventToken, this](
+            ICoreWebView2Environment* sender,
+            ICoreWebView2BrowserProcessExitedEventArgs* args) {
+        COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND kind;
+        CHECK_FAILURE(args->get_BrowserProcessExitKind(&kind));
+
+        // Watch for graceful browser process exit. Let ProcessFailed event
+        // handler take care of failed browser process termination.
+        if (kind == COREWEBVIEW2_BROWSER_PROCESS_EXIT_KIND_NORMAL)
+        {
+          CHECK_FAILURE(
+              m_webViewEnvironment->remove_BrowserProcessExited(browserExitedEventToken));
+          // Release the environment only after the handler is invoked.
+          // Otherwise, there will be no environment to raise the event when
+          // the process exits.
+          m_webViewEnvironment = nullptr;
+          CleanupUserDataFolder();
+        }
+
+        return S_OK;
+    }).Get(),
+    &browserExitedEventToken));
+```
+
+## .NET C#
+```c#
+// URI or other state to save/restore when the WebView is recreated.
+private Uri _uriToRestore;
+
+async void RegisterForNewVersion()
+{
+    // We need to make sure the CoreWebView2 property is not null, so we can get
```

Comment:
Code does not match comment. Comment says we need to make sure the property is not null, but the code just assumes it's not null. Please align. Should apps verify that the property is non-null?

Reply:
`EnsureCoreWebView2Async` will initialize the control and set its `CoreWebView2`. I updated the comment to explain that.